src/CarouselItem.js

```diff
@@ -1,84 +1,62 @@
 import React from 'react';
 import PropTypes from 'prop-types';
 import classNames from 'classnames';
+import Transition from 'react-transition-group/Transition'
 import { mapToCssModules } from './utils';
 import CarouselCaption from './CarouselCaption';
 
 class CarouselItem extends React.Component {
-
   constructor(props) {
-    super(props);
-    this.state = { animation: [] };
-  }
-
-  componentWillUnmount() {
-    clearTimeout(this.willEnterTimeout);
-    clearTimeout(this.willLeaveTimeout);
+    super(props)
+    
+    this.componentWillExit = this.componentWillExit.bind(this);
+    this.componentDidExit = this.componentDidExit.bind(this);
   }
 
-  componentWillAppear(callBack) {
-    this.setState({
-      animation: ['active']
-    });
-    callBack();
-  }
-
-  componentWillEnter(callBack) {
-    const classes = this.context.direction === 'right' ?
-      ['carousel-item-next', 'carousel-item-left'] :
-      ['carousel-item-prev', 'carousel-item-right'];
-    this.setState({
-      animation: classes
-    });
-
-    this.willEnterTimeout = setTimeout(() => {
-      callBack();
-    }, 500);
-  }
-
-  componentDidEnter() {
-    this.setState({
-      animation: ['active']
-    });
-  }
-
-  componentWillLeave(callBack) {
-    const classes = this.context.direction === 'right' ?
-      ['carousel-item-left', 'active'] :
-      ['carousel-item-right', 'active'];
-    this.setState({
-      animation: classes
-    });
-
+  componentWillExit() {
     this.slide.dispatchEvent(new CustomEvent('slide.bs.carousel'));
-
-    this.willLeaveTimeout = setTimeout(() => {
-      callBack();
-    }, 500);
   }
 
-  componentDidLeave() {
-    this.setState({
-      animation: []
-    });
+  componentDidExit() {
     this.slide.dispatchEvent(new CustomEvent('slid.bs.carousel'));
   }
 
-
   render() {
     const { src, altText, children, cssModule } = this.props;
-    const classes = mapToCssModules(classNames(
+    const imgClasses = mapToCssModules(classNames(
         'd-block',
         'img-fluid'
     ), cssModule);
 
-    const itemClasses = mapToCssModules(classNames('carousel-item', ...this.state.animation), cssModule);
 
     return (
-      <div className={itemClasses} ref={(slide) => { this.slide = slide; }}>
-        <img className={classes} src={src} alt={altText} />
-        {children}
-      </div>
+      <Transition
+        in={true}
```

Comment:
Do we want expose Transition props and spread them here? If so, can you also check for the existance of a user provided `onExited` and `onExiting` (via props) and call them in your callbacks (to not clobber the user's callbacks)